File: 739_DailyTemperatures.py
class Solution:
    def dailyTemperatures(self, T: List[int]) -> List[int]:
        stack = []
        res = [0 for i in range(len(T))]
        for i, temp in enumerate(T):
            if stack and stack[-1][0] < temp:
                while stack and stack[-1][0] < temp:
                    res[stack[-1][1]] = i-stack[-1][1]
                    del stack[-1]
            stack.append([temp, i])
        
        return res

File: cpp/739_DailyTemperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& T) {
        vector<pair<int, int>> stack;
        vector<int> results(T.size(), 0);
        for (int i = 0; i < T.size(); ++i) {
            int temp = T[i];
            if (!stack.empty() && temp > stack.back().second) {
                while (!stack.empty() && temp > stack.back().second) {
                    results[stack.back().first] = i - stack.back().first;
                    stack.pop_back();
                }
            }
            stack.push_back({i, temp});
        }
        return results;
    }
};
